dmcontent/govuk_frontend.py

```diff
@@ -0,0 +1,152 @@
+
+    :param question: A Question or QuestionSummary
+    :param data: A dict that may contain the answer for question
+    :param errors: A dict which may contain an error message for the question
+
+    :returns: The name of the macro and the parameters in a tuple, or None
+              if we don't know how to handle this type of question
+    """
+    if question.type == "text":
+        return "govukInput", govuk_input(question, data, errors)
+    else:
+        return None
+
+
+def govuk_input(
+    question: Question, data: Optional[dict] = None, errors: Optional[dict] = None
+) -> dict:
+    """Create govukInput macro parameters from a text question"""
+
+    params = _params(question, data, errors)
+    params["classes"] = "app-text-input--height-compatible"
+
+    return params
+
+
+def _params(
+    question: Question, data: Optional[dict] = None, errors: Optional[dict] = None
+) -> dict:
+    """Common parameters for govuk-frontent components
+
+    The GOV.UK Design System macro library govuk-frontend has a consistent set
+    of parameters that are used across almost all of its component macros.
+
+    This function abstracts out those common parameters to hopefully simplify
+    creating parameter sets for specific macros.
+
+    *This function should however be considered an implementation detail of
+    this module, and you should avoid using it outside of this file (hence the
+    underscore).*
+
+    The parameters handled by this function include:
+
+        - errorMessage (optional)
+        - hint (optional)
+        - id
+        - label
+        - name
+        - value (optional)
+
+    :returns: A dictionary with parameters that are generally useful for
+              govuk-frontend component macros
+    """
+    params = {
+        "id": f"input-{question.id}",
+        "name": question.id,
+    }
+
+    label_text = question.question
+    if question.is_optional:
+        # GOV.UK Design System says
+        # > mark the labels of optional fields with '(optional)'
+        label_text += " (optional)"
+
+    params["label"] = {
+        # Style the label as a page heading, following the
+        # GOV.UK Design System question pages pattern at
+        # https://design-system.service.gov.uk/patterns/question-pages/
+        "classes": "govuk-label--l",
+        "isPageHeading": True,
+
+        "text": label_text,
+    }
+
+    hint_html = Markup()
+    if question.get("question_advice"):
+        # Put the question advice inside the hint, wrapped in a div
+        # We add the class .app-hint--text so the question advice
+        # can be styled like a normal paragraph with the following Sass
+        #
+        #     .app-hint--text {
+        #       @extend %govuk-body--m
+        #     }
+        hint_html += Markup('<div class="app-hint--text">\n')
+        hint_html += escape(question.question_advice)
+        hint_html += Markup("\n</div>")
+        if question.get("hint"):
+            hint_html += "\n"
```

Comment:
My only concern with this is that question_advice can get quite involved ("cultural fit criteria advice", for example, encompasses multiple headings, paragraphs and lists).

The service manual recommends [limiting hint text to about 3 lines max](https://www.gov.uk/service-manual/design/designing-good-questions#how-to-add-help-text).

There's potential for this to be offputting to screenreader users, since inputs are describedBy the hint text.

Can't immediately think of an easy way to place the advice text outside of the hint span that doesn't involve a template/macro.

Reply:
I thought about this, but I'm not sure I understand whether it makes a difference to a screenreader whether the question advice is in a separate div from the hint or not. As far as I can tell hint text is not a semantic thing, so the screenreader is not going to treat it specially from any other text on the page, right?

So if we're putting a huge block of text between the label and the input it's going to get read the same way regardless of the semantic layout, in theory? (I'm not sure about this, so testing with alphagov/digitalmarketplace-briefs-frontend#337 would be much appreciated).

All this is to say that the in practice the way to follow the service manual guidance is to get rid of the question advice, or severely cut it down... which isn't really a technical issue so outside the scope of this PR 😉

Reply:
> ("cultural fit criteria advice", for example, encompasses multiple headings, paragraphs and lists).

😢

Reply:
Inputs have an `aria-describedby` which gets set to the error and hint text if they're present. On Mac VoiceOver, this equates to inputs being read out as [question_label] "Edit Text" [Error text, Hint Text] "You are currently on...", so that's the concern.

However, as you say, that's not really in the scope of this PR, and I'd say this probably qualifies as accessible as is - we're certainly not excluding any information.

Reply:
> Inputs have an aria-describedby which gets set to the error and hint text if they're present. On Mac VoiceOver, this equates to inputs being read out as [question_label] "Edit Text" [Error text, Hint Text] "You are currently on...", so that's the concern.

Hmm, so does that mean if I'm tabbing through the page it will read the hint text twice?

It might be worth taking this to the accessibility team or the design system team. I would like to get this aspect right first time, as much as possible.

Reply:
I would treat questions like `culturalFitCriteria` as an exception, that we should be planning to remove/change the format of anyway. We know those lengthy questions are not a great experience for users, and ideally that huge amount of content would be elsewhere entirely (eg the guidance). So let's not make things overly complicated in order to cater for them.

Reply:
Not an issue if you're tabbing through - you'd only hear it once on the input since you wouldn't tab onto the hint text itself.

It does repeat if you're letting the screenreader read the page for you.

Asked the Design System team:

> I don't think the design system itself has guidance on how long hint text should be, but we've definitely had conversations with people along similar lines where, for example, we've suggested that a long paragraph or bullet point list may be stretching the use of the hint text.
> The way the input component is set up, the input is describedby the hint text, so I imagine long hint text may be offputting for screenreader users

Reply:
> It does repeat if you're letting the screenreader read the page for you.

😬

Reply:
I've spiked having the label separate from the rest of the component, allowing us to drop the question advice in the middle. This seems to work, but the downside is some of the complexity gets pushed back into the apps.

https://github.com/alphagov/digitalmarketplace-briefs-frontend/compare/ldeb-govuk-input...ldeb-govuk-input--separate-label

Thoughts?

Reply:
Of course, this brings up the opposite issue of users tabbing through and not having the advice for context 😫 

As long as the advice is semantic I think that's fine. I think I like this better? This way you won't have the potential for 20 repeats on something like the List Input component.

Reply:
One issue: if the question advice is non-html, it's currently being inserted as a stray bit of text:

![Screenshot 2020-08-11 at 15 56 46](https://user-images.githubusercontent.com/22524634/89913166-4c2c7180-dbeb-11ea-835a-9202d3e1f380.png)

Reply:
I'll look at how to make sure the question advice is always in `p` tag(s). You mentioned the question advice should be semantic, does that mean we should wrap it all in a `div` of some class?

Reply:
Nope, as long as the advice follows standard hierarchy (<h2>, <p>, etc). If any of them get weirder than that we could wrap it, but I think they're all either plain text or html text content.

Reply:
> One issue: if the question advice is non-html, it's currently being inserted as a stray bit of text:

#92 should fix that

Reply:
This all sounds sensible to me. If the content markup in the frameworks is proving difficult to display, we have some flexibility to change it ourselves (changing the content itself is a bit more involved though).

Reply:
I've pushed the spike branch over this branch, so now the label is separate from the input component itself